db/migrate/20201014200514_create_patient_and_jurisdiction_lookups.rb

```diff
@@ -0,0 +1,154 @@
+
+        # Create unique, random, 10 character, url-safe, base-64 strings
+        patient_lookups = submission_tokens.map.with_index do |(id, submission_token), index|
+          { old_submission_token: submission_token, new_submission_token: SecureRandom.urlsafe_base64[0, 10] }
+        end
+
+        # Populate lookup table
+        PatientLookup.import(patient_lookups)
+      end
+
+      # Check for duplicates and generate new submission tokens individually if necessary
+      puts '  Checking for duplicates:'
+      duplicate_tokens = PatientLookup.select(:new_submission_token).group(:new_submission_token).having('COUNT(*) > 1').pluck(:new_submission_token)
+      if duplicate_tokens.empty?
+        puts '    NO DUPLICATES DETECTED'
+      else
+        PatientLookup.where(new_submission_token: duplicate_tokens).find_each do |patient_lookup|
+          puts "    REGENERATING SUBMISSION TOKEN FOR DUPLICATE TOKEN: #{patient_lookup[:new_submission_token]}"
+          # Create a unique, random, 10 character, url-safe, base-64 string
+          new_submission_token = nil
+          loop do
+            new_submission_token = SecureRandom.urlsafe_base64[0, 10]
+          break unless PatientLookup.where('BINARY new_submission_token = ?', new_submission_token).any?
+          end
+
+          # Update patient lookup
+          PatientLookup.where(new_submission_token: patient.submission_token).update(new_submission_token: new_submission_token)
+        end
+      end
+
+      # Update assessment receipts with new submission tokens
+      execute <<-SQL.squish
+        UPDATE assessment_receipts
+        INNER JOIN (
+          SELECT old_submission_token, new_submission_token
+          FROM patient_lookups
+        ) t ON assessment_receipts.submission_token = t.old_submission_token
+        SET assessment_receipts.submission_token = t.new_submission_token
+      SQL
+
+      # Update patients with new submission tokens
+      execute <<-SQL.squish
+        UPDATE patients
+        INNER JOIN (
+          SELECT old_submission_token, new_submission_token
+          FROM patient_lookups
+        ) t ON patients.submission_token = t.old_submission_token
+        SET patients.submission_token = t.new_submission_token
+      SQL
+    end
+
+    # Generate new unique identifiers for existing jurisdictions and populate lookup table
+    puts 'Generating new jurisdiction unique identifiers...'
+    ActiveRecord::Base.transaction do
+      jurisdiction_paths_and_identifiers = Jurisdiction.all.pluck(:id, :path, :unique_identifier)
+
+      # Create new unique identifiers (10 character, url-safe, base-64 strings based on the SHA-256 hash of the jurisdiction paths)
+      puts '  Computing hashes...'
+      new_unique_identifiers = jurisdiction_paths_and_identifiers.collect(&:second).map do |path|
+        Base64::urlsafe_encode64([[Digest::SHA256.hexdigest(path)].pack('H*')].pack('m0'))[0, 10]
```

Comment:
Would be nice to put this generator code in a method on the Jurisdiction model, since it's used twice in these changes already.

Reply:
While that would be possible in the `admin.rake` call, doing that here in the migration would require us modify the query to return each of the ActiveRecord jurisdiction objects, which takes longer than using `.pluck` (the current implementation) because that simply returns arrays of the result.

Although it would be nice to have this code in one place, I don't think it's something that would ever get called outside of `admin.rake` and this migration, and the way it's implemented right now in the migration is optimized for performance.

Reply:
That works for me, thanks!